PEP 458: Add Sections for Signing Metadata and the Formats of Key Objects and Metadata

vladimir-v-diaz opened the pull request:
* Discuss the actual format of the JSON metadata files, and reference the example metadata files included in the "pep-0458-repository" subdirectory available here.

* Explain how the JSON metadata is signed and the signing functions provided by the repository management tool.  Code snippets showing the signing and writing of metadata are provided as well.

* Finally, a section for the password-protected, encrypted key files is added.  In addition to the cryptographic primitives that may be used, the format of the key objects stored in these encrypted key files and metadata is outlined.

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
+--------------------------------------
+
+This section presents the format of the JSON metadata files.  Examples of the
+roles and their formats mentioned here are available for review in the
```
"and their formats mentioned here" seems a bit off. Can you just delete "mentioned here"?

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
+which keys are authorized for all top-level roles, including the root
+role itself.  Revocation and replacement of top-level role keys, including
+for the root role, is done by changing the keys listed for the roles in
+this file.
```
This sentence seems pretty convoluted.

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
+EXPIRES determines when metadata should be considered expired and no longer
+trusted by clients.  Clients MUST NOT trust an expired file.
+
+A ROLE is one of "root", "snapshot", "targets", "timestamp", or "mirrors".
```
438: rather than "is one of" I suggest "may be"

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
+A ROLE is one of "root", "snapshot", "targets", "timestamp", or "mirrors".
+A role for each of "root", "snapshot", "timestamp", and "targets" MUST be
+specified in the key list. The role of "mirror" is optional.  If not
+specified, the mirror list will not need to be signed if mirror lists are
```
line 441: "the mirror list will not need to be signed" add "even" at the end of this phrase

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
+This file can optionally define other roles to which it delegates trust.
+Delegating trust means that the delegated role is trusted for some or all of
+the target files available from the repository. When delegated roles are
+specified, they are specified in a similar way to how the Root role specifies
```
I would change this "they are specified in a similar way to how" to this: "it is done similar to how"

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
+
+The "path_hash_prefixes" list is used to succinctly describe a set of target
+paths. Specifically, each HEX_DIGEST in "path_hash_prefixes" describes a set of
+target paths; therefore, "path_hash_prefixes" is the union over each prefix of
```
"union over" seems odd. should it be "union of"?

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
+}
+
+METAFILES is the same is described for the snapshot.json file.  In the case of
+the timestamp.json file, this will commonly only include a description of the
```
596: I would move "only" to after "include"

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
+  "meta" : METAFILES
+}
+
+METAFILES is the same is described for the snapshot.json file.  In the case of
```
595: "metafiles is the same" ____ (what?) then, use "is" or "as" depending on what works

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -499,16 +701,143 @@ SHOULD also be made available to clients.
+PyCA Cryptography can be added in the future), may override the default number
+of PBKDF2 iterations, and the KDF tweaked to preference.  However, the exact
+cryptographic constructions can be adjusted to include future primitives added
+by the cryptographic libraries supported.
```
735: something is wrong with "supported" - at the very least it is in the wrong place.

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -499,16 +701,143 @@ SHOULD also be made available to clients.
+by the cryptographic libraries supported.
+
+
+Keys objects stored in encrypted key files and in metadata have the format:
```
738: should the first "Keys" be Key?

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -499,16 +701,143 @@ SHOULD also be made available to clients.
+
+  >>> from tuf.repository_tool import *
+
+  # Generate and write an ed25519 key pair.  The private key is saved encrypted.
```
786: "saved encrypted" doesn't seem right. Maybe "encrypted before it is saved"

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -499,16 +701,143 @@ SHOULD also be made available to clients.
+  Enter a password for the encrypted ED25519 key: 
+
+
+The repository tool can use the cryptographic keys that are imported to sign
```
800: do you mean to say "use the cryptographic keys that are imported to sign" or, could you instead say (better): "use the imported cryptographic keys to sign"

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -499,16 +701,143 @@ SHOULD also be made available to clients.
+}
+
+ROLE is a dictionary whose "_type" field describes the role type.
+KEYID is the identifier (64-byte hexstring) of the key signing the ROLE dictionary.
```
change "of the key signing" to "of the key that signs"

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -571,19 +900,19 @@ Project developers expect the distributions they upload to PyPI to be
+metadata and repository files when multiple developers simultaneously update
+the same metadata or distributions.  Without TUF, there are also issues with
+consistency on PyPI, but the problem is more severe with signed metadata that
+MUST keep track in real-time the files available on PyPI.
```
906: revise "keep track in real-time the files" to "keep track, in real-time, of the files"

lvigdor commented on pep-0458-tuf-online-keys.rst:
```diff
@@ -571,19 +900,19 @@ Project developers expect the distributions they upload to PyPI to be
+ensuring consistency of metadata, the client would find itself with a copy of
+*snapshot* that is inconsistent with what is available on PyPI: this situation
+is indistinguishable from arbitrary metadata injected by an attacker.  The
+problem would also occur with mirrors attempting to sync with PyPI.
```
915: change "attempting to" to "that attempt to"

vladimir-v-diaz commented:
@lvigdor 
Thanks you for reviewing this pull request.  Addressed all of your comments here: https://github.com/vladimir-v-diaz/interoperability-peps/commit/c838e229aed78df274e8a867b2877609247e172e
